Add ability to set a password hash for a user. Undef = !.

johnzimm opened the pull request:
I had a use case where I needed to set a password hash for some accounts.

westbywest commented:
I'm curious about the reason for assuming a locked password ('!') for users who have no $password parameter defined.  Why not just let the $password be some value or default undef?  Locked passwords could still be achieved by passing '!' as the $password value.

johnzimm commented:
That would probably be a better approach.  I will update the branch with that modification.

The idea was that if we were managing the account we would want to make sure it either had a known password (or be locked); but not be in a state where we didn't know if it actually had a password set or not.  The preference in our scenario would be to encourage the use of ssh keys.

I'll update so that locking the password is optional as you suggested this afternoon.

westbywest commented:
I actually was about to make a PR of my own that incorporates the changes discussed in this thread:
https://github.com/westbywest/puppet-accounts/commit/63f7d43309e38f0ac5ea52a5b615f901a6bbcae7

Besides the ability to specify a password hash (or leave it undef), I'm also needing the ability to provide relationship metaparameters, i.e. unless, before, require.

johnzimm commented:
I fixed a validation problem with "!" that was pointed out; used single quotes instead.

I also added a second commit that changes the behavior of assuming an undefined password will disable password login with '!' in place of the hash.  It now just leaves the password alone if $password is undef.

Please let me know if you'd prefer these to be squashed to 1 commit.

westbywest commented:
I withdraw any previous requests made in this thread.  Had to switch to another module.

mcanevet commented on manifests/account.pp:
```diff
@@ -43,18 +45,40 @@
-          groups     => $groups,
-          home       => $_home,
-          managehome => true,
-          membership => $groups_membership,
```
I guess you can just add `password => $password` here and don't do the `if $password != undef ... else ... end` block

johnzimm replied:
I guess that does make sense to remove now.  Doing that now.

mcanevet commented:
LGTM
